File: scala/16.scala
object Solution {
    def threeSumClosest(nums: Array[Int], target: Int): Int = {
      val sorted = nums.sorted
      def find(arr: Array[Int], target: Int): Int = {
        var l = 0
        var r = arr.length - 1
        var res = arr(l)+arr(r)
        while (l < r) {
          if ((res - target).abs >= (arr(l) + arr(r) - target).abs) res = arr(l) + arr(r)
          else if (res == target) return res
          if (arr(l) + arr(r) > target) r -= 1
          else l += 1
        }
        res
      }
      val res = for ( i <- 0 until (nums.length - 2) ) yield find(sorted.slice(i+1, nums.length), target - sorted(i)) + sorted(i)
      res.minBy(n => (n-target).abs)
    }
}

File: go/16.go
import "sort"

func threeSumClosest(nums []int, target int) int {
	sort.Ints(nums)

	res := 100000
	for i := range nums {
		if i != 0 && nums[i] == nums[i-1] {
			continue
		}
		for j, k := i+1, len(nums)-1; j < k; {
			sum := nums[i] + nums[j] + nums[k]
			if abs(res-target) > abs(sum-target) {
				res = sum
			}
			if sum < target {
				j++
			} else {
				k--
			}
		}
	}
	return res
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}